decoder.go

```diff
@@ -20,18 +32,52 @@ func frameLineSize(frame *C.AVFrame) *C.int {
 	return (*C.int)(unsafe.Pointer(&frame.linesize[0]))
 }
 
-// h264Decoder is a wrapper around ffmpeg's H264 decoder.
-type h264Decoder struct {
-	codecCtx    *C.AVCodecContext
-	srcFrame    *C.AVFrame
-	swsCtx      *C.struct_SwsContext
-	dstFrame    *C.AVFrame
-	dstFramePtr []uint8
+// getStreamInfo opens a stream URL and retrieves its information
+func getStreamInfo(url string) string {
+	// Register all formats and codecs
+	C.avformat_network_init()
+
+	// Convert Go string to C string
+	cUrl := C.CString(url)
+	defer C.free(unsafe.Pointer(cUrl))
+
+	// Allocate an AVFormatContext.
+	var avFormatCtx *C.AVFormatContext = nil
+
+	// Try to open the stream
+	if C.avformat_open_input(&avFormatCtx, cUrl, nil, nil) != 0 {
+		return "Error: Could not open the stream"
+	}
+	defer C.avformat_close_input(&avFormatCtx)
+
+	// Try to retrieve the stream information
+	if C.avformat_find_stream_info(avFormatCtx, nil) < 0 {
+		return "Error: Could not retrieve the stream information"
+	}
+
+	// Iterate over all streams and check codec ID
+	for i := C.uint(0); i < C.uint(avFormatCtx.nb_streams); i++ {
+		stream := *(**C.AVStream)(unsafe.Pointer(uintptr(unsafe.Pointer(avFormatCtx.streams)) + uintptr(i)*unsafe.Sizeof(*avFormatCtx.streams)))
+		codecParams := stream.codecpar
+
+		codecName := C.GoString(C.avcodec_get_name(codecParams.codec_id))
+
+		if codecParams.codec_id == C.AV_CODEC_ID_H264 {
+			C.avformat_close_input(&avFormatCtx)
```

Comment:
I know this was merged, but maybe double check that this close and the close on line 69 are kosher. Specifically because line 51 uses a defer to call close on the `avFormatCtx`.

I expect double-calling close is fine (i.e: the underlying library copes with that and just early returns, unlike double-free which is a bug).

But if there's a reason lines 66 and 69 need a close before return and line 72 and 76 must not have a close, we should note that for readers.

Reply:
Thank you! Good catch.
Do not need those explicit frees - just rely on the defer statement.